File: Engine/Axis/Source/Engine/Utilities/Serialization/AxHierarchyStreamWriter.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxHierarchyStreamWriter.h"


AxHierarchyStreamWriter::AxHierarchyStreamWriter(AxStream &stream)
{
    this->stream = &stream;
}


AxHierarchyStreamWriter::~AxHierarchyStreamWriter(void)
{
}

void AxHierarchyStreamWriter::BeginChunk(unsigned int chunkId)
{
	this->stream->WriteUInt32(chunkId);
	this->chunkStack.Push(this->stream->position);
	this->stream->WriteUInt64(0);
}

void AxHierarchyStreamWriter::EndChunk()
{
	long long currentPos = this->stream->position;
	this->stream->Seek(this->chunkStack.Pop());
	this->stream->WriteUInt64(currentPos - this->stream->position - 8);
	this->stream->Seek(currentPos);
}

File: Web/Engine/Utilities/Serialization/AxHierarchyStreamWriter.js
/**
 * Creates a new AxHierarchyStreamWriter
 * AxHierarchyStreamWriter utilizes the serialization of data in Axis' heirarchical serialization format
 * @param {AxStream} stream The stream in which to write
 * @constructor
 */
function AxHierarchyStreamWriter(stream)
{
    this.stream = stream;
    this.chunkStack = new AxStack();
}



/**
 * Begins a chunk and writes a chunk header
 * @param {Integer} chunkId Identificator for the chunk
 */
AxHierarchyStreamWriter.prototype.BeginChunk = function(chunkId)
{
    this.stream.WriteUInt32(chunkId);
    this.chunkStack.Push(this.stream.position);
    this.stream.WriteUInt64(0);
};

/**
 * Ends a chunk and finisheds the previously created header fro the chunk.
 * A corresponding call of BeginChunk must have taken place before calling EndChunk
 */
AxHierarchyStreamWriter.prototype.EndChunk = function()
{
    var currentPos = this.stream.position;
    this.stream.Seek(this.chunkStack.Pop());
    this.stream.WriteUInt64(currentPos - this.stream.position - 8);
    this.stream.Seek(currentPos);
};